File: finlib-cpp/src/finlib/portfolio/portfolio.cpp
//
// Created by Andy Pack on 19/06/2025.
//

#include "portfolio.hpp"

namespace finlib {
    Portfolio::Portfolio() {
        handle = finlibrs::portfolio_new();
    }

    void Portfolio::add_asset(std::unique_ptr<PortfolioAsset> asset) {
        finlibrs::portfolio_add_asset(handle, asset->handle);
    }

    void Portfolio::apply_rates_of_change() {
        finlibrs::portfolio_apply_rates_of_change(handle);
    }

    bool Portfolio::valid_sizes() {
        return finlibrs::portfolio_valid_sizes(handle);
    }

    finlibrs::NullableFloat Portfolio::profit_loss() {
        return finlibrs::portfolio_profit_loss(handle);
    }

    size_t Portfolio::size() {
        return finlibrs::portfolio_size(handle);
    }

    // bool Portfolio::valid_weights() {
    //     return finlibrs::portfolio_valid_weights(handle);
    // }

    bool Portfolio::is_valid() {
        return finlibrs::portfolio_is_valid(handle);
    }

    finlibrs::Tuple Portfolio::get_mean_and_std() {
        return finlibrs::portfolio_get_mean_and_std(handle);
    }

    finlibrs::NullableFloat Portfolio::value_at_risk(double confidence, finlibrs::NullableFloat initial_investment) {
        return finlibrs::portfolio_value_at_risk(handle, confidence, initial_investment);
    }

    finlibrs::NullableFloat Portfolio::value_at_risk_afer_time(double confidence,
                                                               finlibrs::NullableFloat initial_investment,
                                                               ptrdiff_t after_time) {
        return finlibrs::portfolio_value_at_risk_afer_time(handle, confidence, initial_investment, after_time);
    }


    finlibrs::NullableFloat Portfolio::value_at_risk_percent(double confidence) {
        return finlibrs::portfolio_value_at_risk_percent(handle, confidence);
    }

    Portfolio::~Portfolio() {
        finlibrs::portfolio_destroy(handle);
    }
}

File: FinLib.NET/FinLib/Portfolio/Portfolio.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using FinLib.Risk;

namespace FinLib.Portfolio;

public class Portfolio: IDisposable, IPayoff<double?>, IValueAtRisk
{
    private readonly unsafe Portfolio_native* _portfolio;
    internal unsafe Portfolio_native* GetPtr() => _portfolio;

    public Portfolio()
    {
        unsafe
        {
            _portfolio = NativeMethods.portfolio_new();
        }
    }

    public void AddAsset(string assetName, double quantity, IEnumerable<double> values)
    {
        unsafe
        {
            var n = Encoding.UTF8.GetBytes(assetName);
            var v = values.ToArray();
            fixed (byte* namePtr = n)
            fixed (double* valuesPtr = v){
                NativeMethods.portfolio_add_asset(_portfolio, NativeMethods.portfolio_asset_new(
                    // portfolioWeight,
                    namePtr, assetName.Length, quantity, valuesPtr, (UIntPtr)v.Length));
            }
        }
    }

    public void AddAsset(PortfolioAsset asset)
    {
        unsafe
        {
            NativeMethods.portfolio_add_asset(_portfolio, asset.GetPtr());
        }
    }

    public bool SizeValid
    {
        get
        {
            unsafe
            {
                return NativeMethods.portfolio_valid_sizes(_portfolio);
            }
        }
    }

    public nuint Count
    {
        get
        {
            unsafe
            {
                return NativeMethods.portfolio_size(_portfolio);
            }
        }
    }

    // public bool ValidWeights()
    // {
    //     unsafe
    //     {
    //         return NativeMethods.portfolio_valid_weights(_portfolio);
    //     }
    // }

    public bool IsValid
    {
        get
        {
            unsafe
            {
                return NativeMethods.portfolio_is_valid(_portfolio);
            }
        }
    }

    public void ApplyRatesOfChange()
    {
        unsafe
        {
            NativeMethods.portfolio_apply_rates_of_change(_portfolio);
        }
    }

    public (double, double)? MeanAndStdDev
    {
        get
        {
            unsafe
            {
                return NativeMethods.portfolio_get_mean_and_std(_portfolio);
            }
        }
    }

    public double? ValueAtRisk(double confidence, double? initialInvestment = null)
    {
        unsafe
        {
            return NativeMethods.portfolio_value_at_risk(_portfolio, confidence, initialInvestment);
        }
    }

    public double? ValueAtRiskPercent(double confidence)
    {
        unsafe
        {
            return NativeMethods.portfolio_value_at_risk_percent(_portfolio, confidence);
        }
    }

    public double? ValueAtRiskAfterTime(double confidence, nint at, double? initialInvestment = null)
    {
        unsafe
        {
            return NativeMethods.portfolio_value_at_risk_afer_time(_portfolio, confidence, initialInvestment, at);
        }
    }

    public double? ProfitLoss
    {
        get
        {
            unsafe
            {
                return NativeMethods.portfolio_profit_loss(_portfolio);
            }
        }
    }

    public double Payoff(double? underlying = null)
    {
        unsafe
        {
            return NativeMethods.portfolio_payoff(_portfolio, underlying);
        }
    }

    private void ReleaseUnmanagedResources()
    {
        unsafe
        {
            NativeMethods.portfolio_destroy(_portfolio);
        }
    }

    public void Dispose()
    {
        ReleaseUnmanagedResources();
        GC.SuppressFinalize(this);
    }

    ~Portfolio()
    {
        ReleaseUnmanagedResources();
    }
}